impls/src/backends/db.rs

```diff
@@ -146,6 +155,179 @@ impl Store {
 		Batch { store: self }
 	}
 
+	/// get an TxLogEntry by parent_key_id (key column) and tx_id, tx_slate_id
+	/// If no input is passed, returns all TxLogEntry transactions.
+	/// If outstanding_only = true then return Received/Sent not confirmed transactions.
+	pub fn get_txs(
+		&self,
+		tx_id: Option<u32>,
+		tx_slate_id: Option<Uuid>,
+		parent_key_id: Option<&Identifier>,
+		outstanding_only: bool,
+	) -> Vec<Serializable> {
+		// initial query (get all TxLogEntry)
+		let mut query = format!(
+			"SELECT * FROM data WHERE prefix = '{}' ",
+			TX_LOG_ENTRY_PREFIX
+		);
+
+		// filter by parent_key_id (key)
+		query = match parent_key_id {
+			Some(key) => format!(
+				"{} AND json_extract(data, '$.root_key_id') = '{}'",
+				query,
+				serde_json::to_string(key).unwrap()
+			),
+			None => query,
+		};
+
+		// filter by tx_id
+		query = match tx_id {
+			Some(id) => format!("{} {} '{}'", query, ID_FILTER, id),
+			None => query,
+		};
+
+		// filter by tx_slate_id
+		query = match tx_slate_id {
+			Some(slate_id) => format!(
+				"{} {} {}",
+				query, "AND json_extract(data, $.slate_id) = ", slate_id
+			),
+			None => query,
+		};
+
+		// get not confirmed AND Received,Sent transactions
+		if outstanding_only {
+			query = format!(
+				"{} {} {} {} {} {} {}",
+				query,
+				"AND confirmed = '0'",
+				"AND status IN ('",
+				TxLogEntryType::TxReceived,
+				"','",
+				TxLogEntryType::TxSent,
+				"')"
+			)
+		};
+
+		self.db
+			.prepare(query)
+			.unwrap()
+			.into_iter()
+			.map(|row| {
+				let row = row.unwrap();
+				ser::deserialize(row.read::<&str, _>("data")).unwrap()
+			})
+			.collect()
+	}
+
+	/// get all OutputData by parent_key_id and tx_id, tx_slate_id
+	/// If no input is passed, returns all OutputData transactions.
+	/// If show_full_history = true then it will return all usable transactions and those already used by other transactions.
+	/// If show_spent = true then it will return all transactions that have already spent.
+	pub fn get_outputs(
+		&self,
+		tx_id: Option<u32>,
+		parent_key_id: Option<&Identifier>,
+		show_full_history: bool,
+		show_spent: bool,
+	) -> Vec<Serializable> {
+		// initial query, get all OutputData
+		let mut query = if show_full_history {
+			format!(
+				"SELECT data WHERE prefix IN ('{}', '{}') ",
+				OUTPUT_PREFIX, OUTPUT_HISTORY_PREFIX,
+			)
+		} else {
+			format!("SELECT data WHERE prefix = '{}' ", OUTPUT_PREFIX)
+		};
+
+		// get transaction key column
+		query = match parent_key_id {
+			Some(key) => format!(
+				"{} {} AND json_extract(data, '$.root_key_id')= '{}'",
+				query,
+				SQLITE_FILTER,
+				serde_json::to_string(key).unwrap()
+			),
```

Comment:
Wrong semantic, you must remove the second {}

```suggestion
				"{} AND json_extract(data, '$.root_key_id') = {}",
				query,
				serde_json::to_string(key).unwrap()
			),
```

Reply:
Also when using the `serde_json::to_string(key).unwrap()` it already returns the quotes, so it must be removed from the query.

To the same fix for any where that uses `serde_json::to_string(key).unwrap()`